### PR TITLE
Show iframe popups on root page

### PR DESCRIPTION
**Pull request by siikamiika**
Replaces https://github.com/FooSoft/yomichan/pull/364.
Fixes https://github.com/FooSoft/yomichan/issues/29.

https://github.com/FooSoft/yomichan/commit/765c1feec9b954c0d0cf79fa2146b0954466c987 does basically the same thing as https://github.com/toasted-nutbread/yomichan/tree/iframe-popups. The difference is that it's compatible with the ID generation in https://github.com/FooSoft/yomichan/pull/411 and doesn't expect the ID to be `'root'`. It should also handle the timing issue.

https://github.com/FooSoft/yomichan/commit/5c216a9416475116ecba191fa6f0e648de94bc83 replaces `frameIdPromise` with `frameId` (https://github.com/FooSoft/yomichan/issues/356).

https://github.com/FooSoft/yomichan/pull/417/commits/ea1da87041fad7eb04f2b70d0bacb26eeb4428bf is the magic, it basically passes an offset starting from 0 to the parent frame checking the offset of the origin frame until it reaches the root frame, and after that it's bounced back using `apiForward` with the accumulated offset and an ID token.

~The iframe lookup logic is duplicated in two places and could be refactored to a new class.~ (done in https://github.com/FooSoft/yomichan/pull/417/commits/b5c3d07ac906dc6905175c6821c193fe6921a899) I also noticed that the offset check adds some latency, but is probably needed for each check because the frames can move between scans. It would be too complex to monitor each iframe for resizes, but a compromise could be to cache the position for some time before checking again.

### CODE REVIEW

**Comment by siikamiika** [comment]
There are some issues positioning the popup when an iframe video is made full screen.

File: ext/fg/js/frame-offset-forwarder.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+        let frameOffsetResolve = null;
+        const frameOffsetPromise = new Promise((resolve) => (frameOffsetResolve = resolve));
+
+        const runtimeMessageCallback = ({action, params}, sender, callback) => {
+            if (action === 'frameOffset' && isObject(params) && params.uniqueId === uniqueId) {
+                chrome.runtime.onMessage.removeListener(runtimeMessageCallback);
+                callback();
+                frameOffsetResolve(params);
+                return false;
```
Comment (toasted-nutbread) [inline ext/fg/js/frame-offset-forwarder.js]: Move `return false` outside of the inner block (just always return false; this won't change the behaviour but will give a consistent return type)

File: ext/fg/js/frame-offset-forwarder.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/* global
+ * apiForward
+ */
+
+class FrameOffsetForwarder {
+    constructor() {
+        this._forwardFrameOffset = window !== window.parent ?
```
Comment (toasted-nutbread) [inline ext/fg/js/frame-offset-forwarder.js]: Prefer to wrap multi-line expressions in parentheses

```js
this._forwardFrameOffset = (
    window !== window.parent ?
    this._forwardFrameOffsetParent.bind(this) :
    this._forwardFrameOffsetOrigin.bind(this)
);

File: ext/fg/js/frame-offset-forwarder.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    _onGetFrameOffset(offset, uniqueId, e) {
+        let sourceFrame = null;
+        for (const frame of document.querySelectorAll('frame, iframe:not(.yomichan-float)')) {
+            if (frame.contentWindow !== e.source) { continue; }
+            sourceFrame = frame;
+            break;
+        }
+        if (sourceFrame === null) { return; }
```
Comment (toasted-nutbread) [inline ext/fg/js/frame-offset-forwarder.js]: May want this to always send a response, even if the frame couldn't be found.

Example case: if the iframe is in a shadow DOM, `document.querySelector` won't return anything, and thus `runtimeMessageCallback` will never be invoked. Repeated calls to `applyOffset` will then result in a memory leak since event callbacks are being added without ever being removed.

File: ext/fg/js/frontend-initialize.js [hunk 4]
```diff
@@ -29,9 +31,32 @@ async function main() {
...
+        const runtimeMessageCallback = ({action, params}, sender, callback) => {
+            if (action === 'rootPopupInformation') {
+                chrome.runtime.onMessage.removeListener(runtimeMessageCallback);
+                callback();
+                rootPopupInformationResolve(params);
+                return false;
+            }
+        };
+        chrome.runtime.onMessage.addListener(runtimeMessageCallback);
+        apiForward('rootPopupInformationGet');
```
Comment (toasted-nutbread) [inline ext/fg/js/frontend-initialize.js]: `rootPopupInformationGet` => `rootPopupRequestInformationBroadcast`

Since it's not actually returning anything

File: ext/fg/js/frontend-initialize.js [hunk 5]
```diff
@@ -29,9 +31,32 @@ async function main() {
     const {id, depth=0, parentFrameId, ignoreNodes, url, proxy=false} = data;
 
     let popup;
-    if (proxy) {
+    if (!proxy && (window !== window.parent)) {
+        let rootPopupInformationResolve;
+        const rootPopupInformationPromise = new Promise((resolve) => (rootPopupInformationResolve = resolve));
+
+        const runtimeMessageCallback = ({action, params}, sender, callback) => {
```
Comment (toasted-nutbread) [inline ext/fg/js/frontend-initialize.js]: We have this pattern in a few places now; we may want to abstract it into something that auto-removes the listener. Something like `yomichan.addListenerTemporary(chrome.runtime.onMessage, callback)`. Not necessary to do it now if you think it will make the PR more complex.

File: ext/fg/js/frontend-initialize.js [hunk 6]
```diff
@@ -29,9 +31,32 @@ async function main() {
...
+                rootPopupInformationResolve(params);
+                return false;
+            }
+        };
+        chrome.runtime.onMessage.addListener(runtimeMessageCallback);
+        apiForward('rootPopupInformationGet');
+
+        const {popupId, frameId} = await rootPopupInformationPromise;
+
+        window._frameOffsetForwarder = new FrameOffsetForwarder();
```
Comment (toasted-nutbread) [inline ext/fg/js/frontend-initialize.js]: Any reason why this is assigned to `window._frameOffsetForwarder`?

File: ext/fg/js/frame-offset-forwarder.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+class FrameOffsetForwarder {
+    constructor() {
+        this._forwardFrameOffset = window !== window.parent ?
+            this._forwardFrameOffsetParent.bind(this) :
+            this._forwardFrameOffsetOrigin.bind(this);
+
+        this._windowMessageHandlers = new Map([
+            ['getFrameOffset', ({offset, uniqueId}, e) => { return this._onGetFrameOffset(offset, uniqueId, e); }]
```
Comment (toasted-nutbread) [inline ext/fg/js/frame-offset-forwarder.js]: Can be simplified

```js
['getFrameOffset', ({offset, uniqueId}, e) => this._onGetFrameOffset(offset, uniqueId, e)]
```

File: ext/fg/js/popup-proxy.js [hunk 8]
```diff
@@ -21,14 +21,14 @@
  */
 
 class PopupProxy {
-    constructor(id, depth, parentId, parentFrameId, url) {
+    constructor(id, depth, parentId, parentFrameId, url, applyFrameOffset=async (x, y) => [x, y]) {
```
Comment (toasted-nutbread) [inline ext/fg/js/popup-proxy.js]: Default `applyFrameOffset` to `null` instead, and then do a `null` check before invoking it later on. This simplifies the default arguments (looks complex) and can also elide an `async` call later on when null.

File: ext/fg/js/frame-offset-forwarder.js (hunk 7 above)
Comment (siikamiika) [reply ext/fg/js/frame-offset-forwarder.js]: This won't actually return anything so I don't know why I added the `return` there. Otherwise this style was copied from `_runtimeMessageHandlers` in `Frontend`. I can still make this one use the short syntax.

Comment (toasted-nutbread) [reply ext/fg/js/frame-offset-forwarder.js]: The reason why that one is bracketed is to make it so that there is no return value, although technically `setVisibleOverride` already doesn't return anything.

File: ext/fg/js/frame-offset-forwarder.js (hunk 1 above)
Comment (siikamiika) [reply ext/fg/js/frame-offset-forwarder.js]: This is from `Popup._listenForDisplayPrepareCompleted`. Was there a reason why the return value was added? I could find in https://developer.chrome.com/apps/runtime that if you return `true` it will keep `sendResponse` (`callback` here) valid until it's called even after returning from the listener.

File: ext/fg/js/frame-offset-forwarder.js (hunk 3 above)
Comment (siikamiika) [reply ext/fg/js/frame-offset-forwarder.js]: I might add a timeout to the promise in `applyOffset`. I'm also considering creating an abstraction for the temporary listener as you mentioned, and it could take timeout as an argument.

File: ext/fg/js/frame-offset-forwarder.js (hunk 1 above)
Comment (toasted-nutbread) [reply ext/fg/js/frame-offset-forwarder.js]: It's probably not needed, particularly because true is never returned. It can be omitted from these functions. Not sure why I added it.

File: ext/fg/js/frontend-initialize.js (hunk 5 above)
Comment (siikamiika) [reply ext/fg/js/frontend-initialize.js]: I'll see if it can be easily done to the places this PR uses, and changed to other places later.

File: ext/fg/js/frontend-initialize.js (hunk 6 above)
Comment (siikamiika) [reply ext/fg/js/frontend-initialize.js]: I was thinking that it would get garbage collected if there wasn't a reference to it anywhere (had this issue with Python earlier), but not sure how JS works. Also the instance created in the `else` block looked funny with just `new FrameOffsetForwarder();`. I could add a method that actually starts the listener to work around the issue while keeping the instance in local scope.

Comment (toasted-nutbread) [reply ext/fg/js/frontend-initialize.js]: It won't get GC'd because there are still refs to it in the event listeners.

```js
window.addEventListener('message', this.onMessage.bind(this), false);
```

This implicitly gives `window` a reference to `this`, since `this.onMessage` is bound.

File: ext/fg/js/frontend-initialize.js (hunk 5 above)
Comment (siikamiika) [reply ext/fg/js/frontend-initialize.js]: https://github.com/FooSoft/yomichan/pull/417/commits/9fdcedb4001a5e86cc7e648c61bd9961a202fa54 implements an abstraction for the pattern that was used here and also in `Popup._listenForDisplayPrepareCompleted`, but that wasn't converted to use the new abstraction yet.

File: ext/fg/js/frame-offset-forwarder.js (hunk 3 above)
Comment (siikamiika) [reply ext/fg/js/frame-offset-forwarder.js]: https://github.com/FooSoft/yomichan/commit/9fdcedb4001a5e86cc7e648c61bd9961a202fa54 adds a timeout that rejects everything that takes more than 30 seconds. The timeouts can be adjusted based on how critical the result is.

File: ext/mixed/js/core.js [hunk 9]
```diff
@@ -288,6 +288,42 @@ const yomichan = (() => {
...
+            if (eventHandler === chrome.runtime.onMessage) {
+                const runtimeMessageCallback = ({action, params}, sender, sendResponse) => {
+                    const cleanupResolve = (value) => {
+                        resolved = true;
+                        eventHandler.removeListener(runtimeMessageCallback);
+                        sendResponse();
+                        resolve(value);
+                    };
+
+                    setTimeout(() => {
```
Comment (toasted-nutbread) [inline ext/mixed/js/core.js]: Store the timer ID and clear the timer in cleanupResolve also (if it hasn't already completed). Clears the timer resource earlier, probably better for memory.

Comment (siikamiika) [reply ext/mixed/js/core.js]: https://github.com/FooSoft/yomichan/pull/417/commits/894c13cc32d70de4a3ca5b78a7ea5fb04b5f791f

**Comment by siikamiika** [comment]
https://github.com/FooSoft/yomichan/pull/417/commits/b64f974d9386ac0d5dd2767ab20caa799a611f14 makes the frame offset calls far less frequent and perceived latency less noticeable by giving the offset a lifetime of 1 second and only calling it synchronously on the first call.

**Comment by siikamiika** [comment]
Now I'm thinking if the cache logic should be moved to `FrameOffsetForwarder`, it could be updated to `PopupProxy` with dependency injection

File: ext/mixed/js/core.js (hunk 9 above)
Comment (siikamiika) [reply ext/mixed/js/core.js]: Btw, I wonder if `clearTimeout` actually needs to check if the timer has been cleared. From what I can gather it's idempotent and you can even call it on ID's that never existed without warnings. Performance wise an `if (bool)` should be faster than a function call (when it's `false`), but it adds a few lines and some indentation.

File: ext/fg/js/frame-offset-forwarder.js (hunk 3 above)
Comment (toasted-nutbread) [reply ext/fg/js/frame-offset-forwarder.js]: This should probably still return some sort of "unknown" indicator and the timeout should be treated as a worst-case scenario. The `getFrameOffset` invocation is likely to happen a lot due to mouse movement, and if the timeout is 5 seconds, a very large message queue can build up.

If an "unknown" indicator is returned, this would also enable the child window to decide to skip the call in the future, since if it's unknown once, it's likely to be unknown forever. Not necessary to implement this now, but I think it's good to have that as possibility.

I think we should also start putting together a HTML page in the test folder with some of the odd edge cases we encounter / can think of, such as the iframe in a shadow DOM and the fullscreen issue.

File: ext/fg/js/popup-proxy.js [hunk 10]
```diff
@@ -51,92 +55,109 @@ class PopupProxy {
...
+    async prepare() {
+        const {id} = await this._invokeHostApi('getOrCreatePopup', {id: this._id, parentId: this._parentId});
+        this._id = id;
+    }
+
     isProxy() {
         return true;
     }
 
+    broadcastRootPopupInformation() {
```
Comment (toasted-nutbread) [inline ext/fg/js/popup-proxy.js]: IMO this implementation should be put into frontend, and frontend should be responsible for checking if it is the root popup. The sending of messages to the entire frame seems like something outside the scope of popups.

File: ext/fg/js/popup-proxy.js [hunk 11]
```diff
@@ -51,92 +55,109 @@ class PopupProxy {
...
-    static _convertDOMRectToJson(domRect) {
-        return {
-            x: domRect.x,
-            y: domRect.y,
-            width: domRect.width,
-            height: domRect.height
-        };
+    async _updateFrameOffset() {
+        const firstRun = this._frameOffsetUpdatedAt === null;
+        const expired = firstRun || this._frameOffsetUpdatedAt < Date.now() - 1000;
```
Comment (toasted-nutbread) [inline ext/fg/js/popup-proxy.js]: Define 1000 as a variable (private member?) so it's doesn't appear to be a magic number.

File: ext/mixed/js/core.js [hunk 12]
```diff
@@ -288,6 +293,45 @@ const yomichan = (() => {
...
+                        timeoutId = window.setTimeout(() => {
+                            timeoutId = null;
+                            reject(new Error(`Listener timed out in ${timeout} ms`));
+                            eventHandler.removeListener(runtimeMessageCallback);
+                        }, timeout);
+                    }
+
+                    const cleanupResolve = (value) => {
+                        if (timeoutId !== null) {
+                            window.clearTimeout(timeoutId);
```
Comment (toasted-nutbread) [inline ext/mixed/js/core.js]: Set timeoutId to null.

File: ext/mixed/js/core.js [hunk 13]
```diff
@@ -288,6 +293,45 @@ const yomichan = (() => {
...
+                reject = _reject;
+            });
+
+            if (eventHandler === chrome.runtime.onMessage) {
+                const runtimeMessageCallback = ({action, params}, sender, sendResponse) => {
+                    let timeoutId = null;
+                    if (timeout !== null) {
+                        timeoutId = window.setTimeout(() => {
+                            timeoutId = null;
+                            reject(new Error(`Listener timed out in ${timeout} ms`));
```
Comment (toasted-nutbread) [inline ext/mixed/js/core.js]: Move `reject` after `eventHandler.removeListener`; looks more return-like that way.

File: ext/mixed/js/core.js [hunk 14]
```diff
@@ -288,6 +293,45 @@ const yomichan = (() => {
...
 
+        getTemporaryListenerResult(eventHandler, userCallback, timeout=null) {
+            let resolve;
+            let reject;
+            const listenerPromise = new Promise((_resolve, _reject) => {
+                resolve = _resolve;
+                reject = _reject;
+            });
+
+            if (eventHandler === chrome.runtime.onMessage) {
```
Comment (toasted-nutbread) [inline ext/mixed/js/core.js]: I would omit this check and just rely on duck-typing. If we include this type, we may as well force eventHandler to be chrome.runtime.onMessage

File: ext/mixed/js/core.js [hunk 15]
```diff
@@ -288,6 +293,45 @@ const yomichan = (() => {
             this.trigger('orphaned', {error});
         }
 
+        getTemporaryListenerResult(eventHandler, userCallback, timeout=null) {
+            let resolve;
+            let reject;
+            const listenerPromise = new Promise((_resolve, _reject) => {
+                resolve = _resolve;
```
Comment (toasted-nutbread) [inline ext/mixed/js/core.js]: Just move everything inside of the promise executor. `listenerPromise` isn't used anywhere other than the return value, so that should be safer (in case of exceptions, etc.), and it avoids the need to capture resolve and reject.

File: ext/mixed/js/core.js (hunk 12 above)
Comment (siikamiika) [reply ext/mixed/js/core.js:317]: I had a comment about this earlier:

> Btw, I wonder if clearTimeout actually needs to check if the timer has been cleared. From what I can gather it's idempotent and you can even call it on ID's that never existed without warnings. Performance wise an if (bool) should be faster than a function call (when it's false), but it adds a few lines and some indentation.

In general I agree that it's good to avoid calling functions with unknown side effects or costly processing multiple times, and I can make it null the ID here to be consistent.

File: ext/mixed/js/core.js (hunk 15 above)
Comment (siikamiika) [reply ext/mixed/js/core.js:300]: Yeah, that would be cleaner. I was too excited about using the pattern I recently learned for passing resolve and reject to other methods in a class :smile:

Comment (siikamiika) [reply ext/mixed/js/core.js]: https://github.com/FooSoft/yomichan/pull/417/commits/5ce02f6d29b663d7d80c30341127b058563a02a9

File: ext/mixed/js/core.js (hunk 14 above)
Comment (siikamiika) [reply ext/mixed/js/core.js]: https://github.com/FooSoft/yomichan/pull/417/commits/34d9ac6ad9bae72c6ce87dc49d30bf5b42908111

File: ext/mixed/js/core.js (hunk 13 above)
Comment (siikamiika) [reply ext/mixed/js/core.js]: https://github.com/FooSoft/yomichan/pull/417/commits/8c32ce7799fac5dc0694ab51ab65d31c3ea76fe1

File: ext/mixed/js/core.js (hunk 12 above)
Comment (siikamiika) [reply ext/mixed/js/core.js:317]: https://github.com/FooSoft/yomichan/commit/8c32ce7799fac5dc0694ab51ab65d31c3ea76fe1

**Comment by siikamiika** [comment]
> This should probably still return some sort of "unknown" indicator and the timeout should be treated as a worst-case scenario. The getFrameOffset invocation is likely to happen a lot due to mouse movement, and if the timeout is 5 seconds, a very large message queue can build up.
> 
> If an "unknown" indicator is returned, this would also enable the child window to decide to skip the call in the future, since if it's unknown once, it's likely to be unknown forever. Not necessary to implement this now, but I think it's good to have that as possibility.

Addressed in https://github.com/FooSoft/yomichan/pull/417/commits/204ae40869f904a34d272204b9ba17d345686752. Updating the timestamp should also partially help in cases when it's always unknown by throttling the checks to one per second.

> I think we should also start putting together a HTML page in the test folder with some of the odd edge cases we encounter / can think of, such as the iframe in a shadow DOM and the fullscreen issue.

I think this is a good idea, I'm currently developing by repeatedly opening a Chrome profile with some of these test cases found in the wild open. There's some delay opening them and I got rate limited more than once :joy:

File: ext/fg/js/frame-offset-forwarder.js (hunk 3 above)
Comment (siikamiika) [reply ext/fg/js/frame-offset-forwarder.js]: https://github.com/FooSoft/yomichan/commit/204ae40869f904a34d272204b9ba17d345686752 also relevant here

Comment (siikamiika) [reply ext/fg/js/frame-offset-forwarder.js]: > This should probably still return some sort of "unknown" indicator and the timeout should be treated as a worst-case scenario. The getFrameOffset invocation is likely to happen a lot due to mouse movement, and if the timeout is 5 seconds, a very large message queue can build up.
> 
> If an "unknown" indicator is returned, this would also enable the child window to decide to skip the call in the future, since if it's unknown once, it's likely to be unknown forever. Not necessary to implement this now, but I think it's good to have that as possibility.

Addressed in https://github.com/FooSoft/yomichan/pull/417/commits/204ae40869f904a34d272204b9ba17d345686752. Updating the timestamp should also partially help in cases when it's always unknown by throttling the checks to one per second.

> I think we should also start putting together a HTML page in the test folder with some of the odd edge cases we encounter / can think of, such as the iframe in a shadow DOM and the fullscreen issue.

I think this is a good idea, I'm currently developing by repeatedly opening a Chrome profile with some of these test cases found in the wild open. There's some delay opening them and I got rate limited more than once :joy:

File: ext/fg/js/popup-proxy.js (hunk 10 above)
Comment (siikamiika) [reply ext/fg/js/popup-proxy.js]: https://github.com/FooSoft/yomichan/pull/417/commits/67c5a8c2fb3fdb0050b7ec448ac8a030c0c9d1fb

File: ext/fg/js/popup-proxy.js (hunk 11 above)
Comment (siikamiika) [reply ext/fg/js/popup-proxy.js]: https://github.com/FooSoft/yomichan/pull/417/commits/320d63cdce7009f1a3733dcdf79f456bfe9131a8

File: ext/mixed/js/core.js (hunk 12 above)
Comment (toasted-nutbread) [reply ext/mixed/js/core.js:317]: Yeah, it technically doesn't need to be checked or null'd since the IDs are never reused. It was mainly a consistency remark.

**Comment by siikamiika** [comment]
It's currently not possible to toggle https://github.com/FooSoft/yomichan/pull/417/commits/9b87a55d533b810028249e34e48551243490a143 without reloading the page. The issue is not limited to this option, and it also affects popup depth as noted in https://github.com/FooSoft/yomichan/issues/426. This PR already includes lots of unrelated changes and supporting it would require major refactoring or at least modifying multiple entry points, so I'll leave that for another pull request.

File: ext/mixed/js/core.js [hunk 16]
```diff
@@ -288,6 +293,42 @@ const yomichan = (() => {
             this.trigger('orphaned', {error});
         }
 
+        getTemporaryListenerResult(eventHandler, userCallback, timeout=null) {
+            if (
```
Comment (toasted-nutbread) [inline ext/mixed/js/core.js:297]:
```js
if (!(
  typeof eventHandler.addListener === 'function' &&
  typeof eventHandler.removeListener === 'function'
))
```
is probably more accurate

File: ext/fg/js/popup-proxy.js [hunk 17]
```diff
@@ -51,92 +55,110 @@ class PopupProxy {
...
+        if (this._frameOffsetPromise === null && !expired) { return; }
+
+        if (this._frameOffsetPromise !== null) {
+            if (firstRun) {
+                await this._frameOffsetPromise;
+            }
+            return;
+        }
+
+        this._frameOffsetPromise = this._getFrameOffset();
```
Comment (toasted-nutbread) [inline ext/fg/js/popup-proxy.js:134]: This could probably be simplified, from line 134-156:

```js
async _updateFrameOffset() {
  // ...

  const promise = this._updateFrameOffsetInner();
  if (firstRun) {
    await promise;
  }
}

async _updateFrameOffsetInner() {
  this._frameOffsetPromise = this._getFrameOffset();
  try {
    const offset = await promise;
    this._frameOffset = offset !== null ? offset : [0, 0];
    this._frameOffsetUpdatedAt = Date.now();
  } catch (e) {
    logError(e);
  } finally {
    this._frameOffsetPromise = null;
  }
}
```

File: ext/fg/js/popup-proxy.js [hunk 18]
```diff
@@ -51,92 +55,110 @@ class PopupProxy {
...
 
-    static _convertDOMRectToJson(domRect) {
-        return {
-            x: domRect.x,
-            y: domRect.y,
-            width: domRect.width,
-            height: domRect.height
+    async _updateFrameOffset() {
+        const firstRun = this._frameOffsetUpdatedAt === null;
+        const expired = firstRun || this._frameOffsetUpdatedAt < Date.now() - PopupProxy._frameOffsetExpireTimeout;
```
Comment (toasted-nutbread) [inline ext/fg/js/popup-proxy.js:124]: There are two `Date.now` calls which may have different values; should the value be stored in a variable at the start of the function? Or is it intended that it will be updated after the async calls?

Comment (siikamiika) [reply ext/fg/js/popup-proxy.js:124]: It doesn't matter too much actually because both the time of initiation and getting the results are a bit off. I can make both use the time of initiation for consistency.

File: ext/fg/js/popup-proxy.js (hunk 17 above)
Comment (siikamiika) [reply ext/fg/js/popup-proxy.js:134]: You're right, that's a lot cleaner. `Promise`s all the way

File: ext/fg/js/popup-proxy.js (hunk 18 above)
Comment (siikamiika) [reply ext/fg/js/popup-proxy.js]: https://github.com/FooSoft/yomichan/pull/417/commits/883759b9b7f1d3b5ddb8d3f17a29841c83f725dd

File: ext/fg/js/popup-proxy.js (hunk 17 above)
Comment (siikamiika) [reply ext/fg/js/popup-proxy.js]: https://github.com/FooSoft/yomichan/pull/417/commits/883759b9b7f1d3b5ddb8d3f17a29841c83f725dd

File: ext/mixed/js/core.js (hunk 16 above)
Comment (siikamiika) [reply ext/mixed/js/core.js]: https://github.com/FooSoft/yomichan/pull/417/commits/ff6b79fdc7ef3fd7be4384dec570138e0ad27012

Comment (toasted-nutbread) [reply ext/mixed/js/core.js]: Should be && not || now

**Comment by toasted-nutbread** [comment]
I think this is the case, but just to confirm: without changing any settings, the default behaviour of the extension should currently be the same?

**Comment by toasted-nutbread** [comment]
Looks like my PR merges caused a few merge conflicts, but they should be easy to fix since it's all options related.

**Comment by siikamiika** [comment]
> I think this is the case, but just to confirm: without changing any settings, the default behaviour of the extension should currently be the same?

Apart from some changes to PopupProxy prepare, you're right. The feature itself will be disabled by default and behind advanced settings.

> Looks like my PR merges caused a few merge conflicts, but they should be easy to fix since it's all options related.

No worries, I'll fix it now.

File: ext/mixed/js/core.js (hunk 16 above)
Comment (siikamiika) [reply ext/mixed/js/core.js]: Sorry, that was careless. I tested it and assumed everything was fine because OR short circuited on the first check. https://github.com/FooSoft/yomichan/pull/417/commits/22a97d916fc6ecab1200b0ffea18cf2d5c9923d4
